c-dependencies/js-compute-runtime/builtins/subtle-crypto.cpp

```diff
@@ -164,8 +164,109 @@ bool SubtleCrypto::importKey(JSContext *cx, unsigned argc, JS::Value *vp) {
   return true;
 }
 
+// https://w3c.github.io/webcrypto/#SubtleCrypto-method-sign
+bool SubtleCrypto::sign(JSContext *cx, unsigned argc, JS::Value *vp) {
+  MOZ_ASSERT(cx);
+  MOZ_ASSERT(vp);
+  JS::CallArgs args = CallArgsFromVp(argc, vp);
+
+  if (!args.requireAtLeast(cx, "SubtleCrypto.sign", 3)) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (!check_receiver(cx, args.thisv(), "SubtleCrypto.sign")) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  // 1. Let algorithm and key be the algorithm and key parameters passed to the sign() method,
+  // respectively.
+  auto algorithm = args.get(0);
+  JS::RootedObject key(cx);
+  {
+    auto key_arg = args.get(1);
+    if (!key_arg.isObject()) {
+      JS_ReportErrorLatin1(cx, "parameter 2 is not of type 'CryptoKey'");
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+    key.set(&key_arg.toObject());
+    if (!CryptoKey::is_instance(key)) {
+      JS_ReportErrorLatin1(cx, "parameter 2 is not of type 'CryptoKey'");
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+  }
+
+  // 2. Let data be the result of getting a copy of the bytes held by the data parameter passed to
+  // the sign() method.
+  std::span<uint8_t> data;
+  {
+    std::optional<std::span<uint8_t>> dataOptional =
+        value_to_buffer(cx, args.get(2), "SubtleCrypto.sign: data");
+    if (!dataOptional.has_value()) {
+      // value_to_buffer would have already created a JS exception so we don't need to create one
+      // ourselves.
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+    data = dataOptional.value();
+  }
+
+  // 3. Let normalizedAlgorithm be the result of normalizing an algorithm, with alg set to algorithm
+  // and op set to "sign".
+  auto normalizedAlgorithm = CryptoAlgorithmSignVerify::normalize(cx, algorithm);
+  // 4. If an error occurred, return a Promise rejected with normalizedAlgorithm.
+  if (!normalizedAlgorithm) {
+    // TODO Rename error to NotSupportedError
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  // 5. Let promise be a new Promise.
+  JS::RootedObject promise(cx, JS::NewPromiseObject(cx, nullptr));
+  if (!promise) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  // 6. Return promise and perform the remaining steps in parallel.
```

Comment:
I think we need to hold from from adding to the async event queue. The reason I think this is because "the remaining steps" are implemented within the JS runtime, not within the host, and we don't have wasi-threads yet, which means there is no way for "the remaining steps" to happen in parallel to other execution in the runtime.

When/If wasi-threads is available, we can update our async eventing system to include them and then update subtle-crypto to make use of the async event queue. I could create an issue to keep track of this.

What do you think?

Reply:
That seems reasonable to me! I wasn't sure if there would end up being an observable difference between deferring the rest of the work or returning a resolved promise, but if there's not let's just leave this as-is.

Reply:
There no observed difference 👍